AutoCollection/HeartBeat.ts

```diff
@@ -61,16 +73,88 @@ class HeartBeat {
+                waiting = true;
+                this._getAzureComputeMetadata(config, () => {
+                    if (this._isVM && Object.keys(this._vmData).length > 0) { // VM
+                        properties["azInst_vmId"] = this._vmData["vmId"] || "";
+                        properties["azInst_subscriptionId"] = this._vmData["subscriptionId"] || "";
+                        properties["azInst_osType"] = this._vmData["osType"] || "";
+                        this._azInst_vmId = this._vmData["vmId"] || "";
+                        this._azInst_subscriptionId = this._vmData["subscriptionId"] || "";
+                        this._azInst_osType = this._vmData["osType"] || "";
+                    }
+                    this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+                    callback();
+                });
+            } else if (this._isVM) {
+                properties["azInst_vmId"] = this._azInst_vmId;
+                properties["azInst_subscriptionId"] = this._azInst_subscriptionId;
+                properties["azInst_osType"] = this._azInst_osType;
+            }
+        }
+        if (!waiting) {
+            this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+            callback();
+        }
     }
 
     public dispose() {
         HeartBeat.INSTANCE = null;
         this.enable(false);
         this._isInitialized = false;
     }
+
+    private _getAzureComputeMetadata(config: Config, callback: () => void) {
+        const metadataRequestUrl = `${AIMS_URI}?${AIMS_API_VERSION}&${AIMS_FORMAT}`;
+        const requestOptions = {
+            method: 'GET',
+            [AutoCollectHttpDependencies.disableCollectionRequestOption]: true,
+            headers: {
+                "Metadata": "True",
+            }
+        };
+
+        const req = Util.makeRequest(config, metadataRequestUrl, requestOptions, (res) => {
+            if (res.statusCode === 200) {
+                // Success; VM
+                this._isVM = true;
+                let virtualMachineData = "";
+                res.on('data', (data: any) => {
+                    virtualMachineData += data;
+                });
+                res.on('end', () => {
+                    this._vmData = this._isJSON(virtualMachineData) ? JSON.parse(virtualMachineData) : {};
+                    callback();
+                });
+            } else if (res.statusCode >= 400 && res.statusCode < 500) {
+                // Not found; Not in VM; Do not try again.
+                this._isVM = false;
+                callback();
+            } else {
+                // else Retry on next heartbeat metrics call
+                callback();
+            }
+        });
+        if (req) {
+            req.on('error', (error: Error) => {
+                // Unable to contact endpoint.
+                // Do nothing for now.
+                this._isVM = false;
```

Comment:
How come we don't attempt to retry if we get a different exception besides not being in a vm?

Reply:
So is it only confirm it's not vm when connection error happens? if I understand your [implementation ](https://github.com/census-instrumentation/opencensus-python/blob/67502e512ab073382b4fd12580c3b4534968129b/contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/heartbeat_metrics/heartbeat.py#L105)correctly. But @markwolff I don't think there's a way to check it like python does? My assumption is 'error' === can't connect === not vm

Reply:
I don't think you should make an assumption. I suggest you should try out azuremetadataservice yourself, within a VM, outside a VM, within a VM but exception happens, outside a VM and exception happens and see how it behaves.

Reply:
yes i did. outside a vm is Error: connect ENETUNREACH, it's the connection error, in a vm it returns JSON.

Reply:
And I think the statuscode scenarios is similar to what you check for exception types. 

`1xx informational response – the request was received, continuing process` - retry
`2xx successful – the request was successfully received, understood, and accepted` - vm true
`3xx redirection – further action needs to be taken in order to complete the request` - retry
`4xx client error – the request contains bad syntax or cannot be fulfilled` - vm false
`5xx server error – the server failed to fulfil an apparently valid request` - retry

Reply:
Right, but when you are not in a VM, and it throws Error: connect ENETUNREACH, does that mean it still returns a RESPONSE with a status code? If it does not, you can't rely on `else if (res.statusCode >= 400 && res.statusCode < 500) {` to be doing the check correct?

Reply:
discussed offline, update in next commit